metaflow/plugins/kfp/kfp.py

```diff
@@ -358,33 +368,34 @@ def build_kfp_dag(node: DAGNode, context, index=None):
 
                         visited[step].after(visited[node.name])
 
-            build_kfp_dag(self.graph["start"], {})
+            build_kfp_dag(self.graph["start"], context="")
 
         return kfp_pipeline_from_flow
 
 
 def step_op_func(
-    cmd_template, kfp_run_id, contexts, index=None
+    datastore_root: str, cmd_template: str, kfp_run_id: str, context, index=None
 ) -> NamedTuple("context", [("task_out_dict", dict), ("split_indexes", list)]):
     """
     Function used to create a KFP container op that corresponds to a single step in the flow.
     """
     import os
-    from io import StringIO
+    import json
     import tempfile
+    from io import StringIO
     from subprocess import Popen, PIPE, STDOUT
-    import json
     from typing import NamedTuple
 
     print("----")
     print("context")
-    print(type(contexts))
-    print(contexts)
+    print(context)
     print("----")
-    context_dict = json.loads(contexts)
+    context_dict = json.loads("{}" if context == "" else context)
```

Comment:
Maybe I missed this in earlier PRs but the `context` here is injected at runtime by KFP?

Just asking as it looks like we also set this at compile time (I think) during `build_kfp_dag` which I'm guessing cannot actually resolve the `context_dict.get("task_id", "")`.

Or is it actually executed both at compile and run time?

Reply:
True, context is being passed along to get the parent taskId at runtime, 

True this [code](https://github.com/zillow/metaflow/blob/8c2913661d7eb5bb5417500f9da4bca83bd3dcd4/metaflow/plugins/kfp/kfp.py#L161-L207) will deprecate soon, especially the [current_task_id](https://github.com/zillow/metaflow/blob/8c2913661d7eb5bb5417500f9da4bca83bd3dcd4/metaflow/plugins/kfp/kfp.py#L161-L161).  

TaskIds will be defined at runtime by KFP/Argo.  This design of knowing the taskIds at compile time doesn't work with foreach, where the number of splits is only known at runtime.